design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,202 @@
+likely not the best option since it may be necessary to select individual items from the queue at
+any position and not merely grab the top/bottom item.
+
+Also keep a record of the number of times each OSD has be queued to determine when Rook should stop
+attempting to update the OSD deployment. This can be attached to each OSD in the update queue
+via a struct, or it could be a map structure that can be easily looked up. Attaching via a struct
+would also allow easily attaching additional items (like a list of errors, for example) to OSDs in
+the future.
+
+`kubectl describe deployment rook-ceph-osd-0` on a test cluster yields ~8kB of text. If we assume
+that an OSD Deployment that is more complicated contains 10kB of data, listing all OSD Deployments
+for a cluster with 1000 OSDs would use approximately 10MB of RAM. An extremely large cluster with
+10000 OSDs would use approximately 100MB of RAM. This is likely acceptable for any modern system.
+There is an option in the future to build the update queue by listing OSDs in a paginated manner and
+building the queue out of the minimal set of information needed to perform the updates to reduce
+this usage.
+
+#### Get a batch of OSDs from the update queue
+
+Expect to use topology labels on deployments.
+
+if a batch query reports that it's safe to update OSDs, update all deployments
+  corresponding to batch of OSDs in goroutines and wait for results.
+
+Workflow for when Ceph reports that it isn't safe to update a batch:
+1. query individual OSDs in the batch to see if any individual OSDs are hindering the batch update
+1. requeue any individual OSDs that can't be updated
+1. retry the batch query but without including in the query individual OSDs that can't be updated
+1. if the batch still can't be updated (or if there are no individual OSDs that can't be updated),
+halve the batch and try the batch query again with each half
+1. optionally try to reduce the batch into 1/4ths, 1/8ths, 1/16ths,... and retry
+  > **Q** is this worth it? Maybe just 1/4ths, if applicable?
+1. finally, try individually querying OSDs to update
+1. make sure to requeue any OSDs that weren't updated due to queries
+
+Keep track of OSDs for which queries fail and the number of times they have failed. Once queries for
+an OSD fail more than 3 times, report that the OSD could not be updated. Do not fail the operator.
+
+When requeuing, build a new queue using OSDs that can't be stopped right now in order to not get
+stuck looping on the same osds repeatedly in the first queue.
+
+
+### Configuration
+
+#### Additions to `CephCluster` CRD
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+# ...
+spec:
+  # ...
+  # keep these here for backwards compatibility
+  skipUpgradeChecks:
+  continueUpgradeAfterChecksEvenIfNotHealthy:
+  removeOSDsIfOutAndSafeToDestroy:
+  #
+  # Specify policies related to updating the Ceph cluster and its components
+  updatePolicy:
```

Comment:
To keep it simple, I wonder if we would only need one setting in the cluster CR. 
- osd.maxInParallel: 3

To play it safe, we could default to 3. If 0 (or not set), we could update all in a failure domain.

We wouldn't need a topology setting since we could automatically move up the CRUSH tree as described in the other comment.

Reply:
I think this makes sense. KISS is usually a good strategy. I want to make sure we plan for future expansion of options, which I suspect will happen over time, but starting simply sounds great to me.

Reply:
In this case, it could be ``ceph osd ok-to-stop $id --max 3`` (where basically ``max`` defaults to 1 to capture legacy behavior)...?

Reply:
I think this makes sense 👍